WIP: tests: helper: Add tests for set_count.

kaustubh-nair opened the pull request:
- [x] Test streams
- [x] Test pms
- [x] Test self-pms
- [x] Test group pms
- [x] ~~Test messages sent by user~~ ( `set_count` not called)
- [x] Test muted streams
- [x] Test muted topics
- [ ] ~~Test topics whose view is open~~ ( UI test, fix in different PR )
- [x] Test screen updated
- [ ] Test mentions

kaustubh-nair commented:
This is an important stepping stone to fixing #535.
Handling read events will require modifications to `set_count`. Since, `set_count` is not being tested at all, I'm adding tests for the same. It might be difficult to cover every faucet of `set_counts`, but the goal is to test atleast the major features, so that significant bugs do not arise.

kaustubh-nair commented:
There's will be a little bit of duplication across the tests for setting up and initializing variables. I'm wondering if it would be viable to abstract out the setup, or would it make the tests too complex?

kaustubh-nair commented on tests/helper/test_helper.py:
```diff
@@ -1,13 +1,85 @@
+    (-1, [15], 4, 3, {200: 2, 201: 1}),   # self-pm
+    (1, [16, 10], 6, 5, {200: 4, 201: 1}),
+    (1, [], 4, 3, {200: 2, 201: 1}),
+    # (1, [15], 4, 3, {200: 2, 201: 1}),   # self-pm  # FIXME
```
The test fails when a pm sent by the user is to be marked as unread. I don't know if such a case will ever arise, but `set_count` should probably handle it anyway.

sumanthvrao replied:
I vaguely remember discovering the same fact a while ago. I think the reason I concluded was that the server does not set the unread for self PM's hence there is no point setting the counts from ZT side

kaustubh-nair replied:
Okay, I'll remove this then

kaustubh-nair commented:
@neiljp @sumanthvrao, if you have time to give this a review please let me know if these tests look okay. :) 
And also the extent to which I should be testing `set_count`, so I can go ahead with the rest of the tests?

sumanthvrao commented:
This looks like a great start @kaustubh-nair .  If possible we can try reusing some fixtures in `conftest.py` which may help us avoid the duplication (not entirely I suppose).

The list you have mentioned above seems comprehensive enough! Let's go ahead with that.

kaustubh-nair commented:
The fixtures in `conftest.py` don't contain enough messages to properly test the counts, so I ended up adding new ones

kaustubh-nair committed:
tests: helper: Test set_count for muted topics.

Add test for verifying unread_counts when marking muted topics
read/unread. The 'all_msg' attribute in unread counts does not change
while changing the read status of such messages.

kaustubh-nair committed:
tests: helper: Test screen updated on setting unread counts.

Ensure controller.update_screen is called while setting unread counts.

kaustubh-nair commented:
@sumanthvrao @neiljp This is ready for an initial review. I do realize there is a LOT of duplication going on here. I will push some refactor commits sometime later tomorrow.

kaustubh-nair committed:
tests: helper: Test set_count for stream messages.

Add test for verifying unread_counts when marking streams as
read/unread.

Added new index fixture that contains a large number of messages.

kaustubh-nair committed:
tests: helper: Test set_count for private messages.

Add test for verifying unread counts when marking self-pm and one-on-one
pm as read/unread.

Add new pm entries to 'index_multiple_messages' fixture.

kaustubh-nair committed:
tests: helper: Test set_count for group private messages.

Add test for verifying unread counts when marking group-pm
as read/unread.

Add new pm entries to 'index_multiple_messages' fixture.

kaustubh-nair committed:
tests: helper: Test set_count for muted streams.

Add test for verifying unread_counts when marking muted streams
read/unread. The 'all_msg' attribute in unread counts does not change
while changing the read status of such messages.
